File: Lab_1/main_oop.py
import math
import sys

class Sq_Roots:
    """Класс коэффициентов"""
    def __init__(self):
        self.A = 0.0
        self.B = 0.0
        self.C = 0.0
        self.root_list = set()

    def check_root(self,ind):
        """Проверка коэфов"""
        try:
            global coef
            try:
                coef = float(sys.argv[ind])
            except ValueError:
                print("Ошибка. Попробуйте еще раз")
        except:
            print("Введите коэффициент {} :".format(ind) )
            try:
                coef = input()
                # Проверка А на ноль(иначе впоследствие ошибка - делениe на ноль)
                if float(coef) == 0.0 and ind == 1:
                    print("Коэффициент 1 равен 0. Так не пойдет")
                    self.check_root(1)
            except ValueError:
                self.check_root(ind)
        return float(coef)

    def get_roots(self):
        """Получение значений коэфов"""
        self.A = self.check_root(1)
        self.B = self.check_root(2)
        self.C = self.check_root(3)

    def calculation(self):
        """Подсчет корней"""
        #Дискриминант
        all_roots = set()
        D = self.B**2 - 4*self.A*self.C
        if (D>=0):
            self.root_list.add( (-self.B + math.sqrt(D))/ (2*self.A) )
            self.root_list.add( (-self.B - math.sqrt(D))/ (2*self.A) )
        for r in self.root_list:
            if r >= 0:
                all_roots.add(math.sqrt(r))
                all_roots.add(- math.sqrt(r))
        self.root_list = all_roots

    def print_ans(self):
        """Вывод ответов"""
        if len(self.root_list)==0:
            print("Нет корней, дискриминант меньше нуля :(")
            return
        print("Корни:")
        for e in self.root_list:
            print(e, " ")

def main():
    """Основная функция"""
    sr = Sq_Roots()
    sr.get_roots()
    sr.calculation()
    sr.print_ans()

if __name__ == "__main__":
    main()

File: Lab_1-2-go/main_oop.go
package main

import (
	"bufio"   // для буферизованного ввода/вывода
	"fmt"     // для форматированного ввода/вывода
	"math"    // для математических функций
	"os"      // для взаимодействия с операционной системой
	"strconv" // для преобразования строк
	"strings" // для работы со строками
)

type Coefficients struct {
	A, B, C float64
}

// Метод для получения коэффициентов
func (c *Coefficients) get_roots() {
	for i := 0; i < 3; i++ {
		switch i {
		case 0:
			c.A = check_coefficient(i + 1)
		case 1:
			c.B = check_coefficient(i + 1)
		case 2:
			c.C = check_coefficient(i + 1)
		}
	}
}

// Метод для вычисления корней уравнения
func (c *Coefficients) calculate() []float64 {
	D := math.Pow(c.B, 2) - 4*c.A*c.C
	root_list := make([]float64, 0)
	all_roots := make([]float64, 0)
	if D >= 0.0 {
		root_list = append(root_list, (-c.B+math.Sqrt(D))/(2.0*c.A))
		root_list = append(root_list, (-c.B-math.Sqrt(D))/(2.0*c.A))
	}
	for _, r := range root_list {
		if r >= 0 {
			all_roots = append(all_roots, math.Sqrt(r))
			all_roots = append(all_roots, -math.Sqrt(r))
		}
	}
	return all_roots
}

func main() {
	var coef Coefficients
	coef.get_roots()               // Получаем коэффициенты
	print_answer(coef.calculate()) // Выводим корни уравнения
}

// Функция для вывода корней уравнения
func print_answer(root_list []float64) {
	if len(root_list) == 0 {
		fmt.Println("Нет корней, дискриминант меньше нуля :(")
	} else {
		fmt.Println("Корни:")
		for _, e := range root_list {
			fmt.Println(e, " ")
		}
	}
}

// Функция для проверки корректности введенных коэффициентов
func check_coefficient(ind int) float64 {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Printf("Введите коэффициент %d: ", ind)
		input, _ := reader.ReadString('\n')
		coef, err := strconv.ParseFloat(strings.TrimSpace(input), 64)
		if err != nil {
			fmt.Println("Ошибка. Попробуйте еще раз")
			continue
		}
		if coef == 0.0 && ind == 1 {
			fmt.Println("Коэффициент 1 равен 0. Так не пойдет")
			continue
		}
		return coef
	}
}
